### ISSUE TITLE
Database entities should not use services internally

### ISSUE THREAD

**Issue by ybnd** [DSpace/DSpace#9027]
**Describe the bug**
A recent addition to `Community` and `Collection` can cause NPEs because it depends on an attribute that may not have been set yet ([here](https://github.com/DSpace/DSpace/pull/8843/files#diff-8b4c093a71a836ef5e2406f79542d06d2e64b81f6019a9efb704302f3176fad6R347) and [here](https://github.com/DSpace/DSpace/pull/8843/files#diff-8b4c093a71a836ef5e2406f79542d06d2e64b81f6019a9efb704302f3176fad6R347)).
This issue can occur right after starting Tomcat, if `countArchivedItems` is called _before_ other methods that use `getCommunityService` or `getCollectionService`.

The quick solution would be to use these methods instead of the attributes themselves, but we could also take this opportunity to get rid of this pattern altogether.

All of our `DSpaceObject` subclasses contain at least one similar method that requires a service.
As far as I can tell this originated as a workaround during the initial service refactor to support legacy functionality, but ideally we should move these methods to the services themselves.

This concerns the following classes & methods:
* `DSpaceObject.getName`
* `Bitstream`
  * `setName`
  * `getSource`
  * `setSource`
  * `getDescription`
  * `setDescription`
  * `getUserFormatDescription`
  * `setUserFormatDescription`
  * `getFormatDescription`
  * `getFormat`
  * `setFormat`
  * `setAcceptanceDate`
* `BitstreamFormat`
  * `setShorthandDescription`
  * `setSupportLevel`
* `Bundle`
  * `setName`
* `Collection`
  * `getWorkflowStep1`
  * `getWorkflowStep2`
  * `getWorkflowStep3`
  * `getLicenseCollection`
  * `setLicense`
  * `setWorkflowGroup`
  * `countArchivedItems`
* `Community`
  * `countArchivedItems`
* `Item`
  * `getBundles` (uses `Bundle.getName`)
* `Site`
* `EPerson`
  * `equals`
  * `hashCode`
  * `getLanguage`
  * `setLanguage`
  * `getFullName`
  * `getFirstName`
  * `setFirstName`
  * `getLastName`
  * `setLastName`

**Expected behavior**
"Database object classes" should not contain references to services; existing methods should be moved to services, all usages should be refactored to use the service instead.

**Related work**
https://github.com/DSpace/DSpace/pull/8843

**Comment by tdonohue** [comment]
Agreed, this would be nice to have.  It'd definitely require a lot of effort to achieve, but I do agree it'd result in cleaner code.  Pulling this to the 8.0 board, but marking as low priority for now.  Nonetheless, I'd support this being done if a volunteer steps forward.

**Comment by mwoodiupui** [comment]
Collection#getWorkflowStep[123] should disappear altogether, as workflow steps don't work that way anymore.

**Comment by ybnd** [comment]
I'm going to take a stab at it (but it will probably take a while still)
